From 95bbc2759f9c9f8f93a233f99d49e4a5d0d87705 Mon Sep 17 00:00:00 2001
From: Jiani Liu <quic_jianil@quicinc.com>
Date: Thu, 6 Apr 2023 17:22:28 +0800
Subject: [PATCH] ath9k: fix ANI crash

Start ANI through debugfs before ath9k is started, will
cause kernel crash.

CPU 0 Unable to handle kernel paging request at virtual
address 00000000, epc == 82f936fc, ra == 82f91788
Oops[#1]:
CPU: 0 PID: 0 Comm: swapper Not tainted 4.4.60 #246
task: 8045ee98 ti: 80458000 task.ti: 80458000
$ 0 : 00000000 804b0000 82f02320 00009860
$ 4 : 82f1c010 00000000 00000500 00000001
$ 8 : 00008098 80459cc8 00008090 00008094
$12 : 00000013 772ad350 00000000 00000000
$16 : 82f1c010 00000001 fffcaf4a 000003e8
$20 : 00000000 00000000 00000001 00000001
$24 : 00000000 80213cc4
$28 : 80458000 80459c70 00000004 82f91788
Hi : 005f5dd1
Lo : ffffffff
epc : 82f936fc ath9k_hw_getnf+0x2c/0x31c [ath9k_hw]
ra : 82f91788 ar9002_hw_attach_calib_ops+0xa8c/0x2898 [ath9k_hw]
Status: 1000fc03 KERNEL EXL IE
Cause : 00800008 (ExcCode 02)
BadVA : 00000000
PrId : 00019374 (MIPS 24Kc)
...
Call Trace:
[<82f936fc>] ath9k_hw_getnf+0x2c/0x31c [ath9k_hw]
[<82f91788>] ar9002_hw_attach_cali...

Signed-off-by: Miaoqing Pan <miaoqing@codeaurora.org>
Signed-off-by: Jiani Liu <quic_jianil@quicinc.com>

diff --git a/drivers/net/wireless/ath/ath9k/debug.c b/drivers/net/wireless/ath/ath9k/debug.c
index ea47acd..bb8f28a 100644
--- a/drivers/net/wireless/ath/ath9k/debug.c
+++ b/drivers/net/wireless/ath/ath9k/debug.c
@@ -266,7 +266,8 @@ static ssize_t write_file_ani(struct file *file,
 		clear_bit(ATH_OP_ANI_RUN, &common->op_flags);
 		ath_stop_ani(sc);
 	} else {
-		ath_check_ani(sc);
+		if (!test_bit(ATH_OP_INVALID, &common->op_flags))
+			ath_check_ani(sc);
 	}
 
 	return count;
-- 
2.7.4

